### PR TITLE
Implement Premade products

### PR DESCRIPTION
**Pull request by jfquevedol2198**
image

### CODE REVIEW

File: generator/src/routes/+page.svelte [hunk 1]
```diff
@@ -39,12 +42,14 @@
 
   async function onGenerate(event: SubmitEvent) {
     event.preventDefault()
+    clearError()
+
     status = Status.GENERATING
     const formEl = event.target as HTMLFormElement
     const formData = new FormData(formEl)
+    if (!product) return
```
Comment (lietu) [inline generator/src/routes/+page.svelte:50]:
```suggestion
    if (!product) {
      return
    }
```

File: generator/src/routes/+page.svelte [hunk 2]
```diff
@@ -98,12 +103,29 @@
     document.body.removeChild(link)
   }
 
-  function onChangeValidSignature(event: Event) {
-    const target = event.target as HTMLInputElement
-    if (target.checked) {
+  function clearError() {
+    error = null
+  }
+  function onChangeValidSignature() {
```
Comment (lietu) [inline generator/src/routes/+page.svelte:109]:
```suggestion
  }
  
  function onChangeValidSignature() {
```

File: generator/src/routes/+page.svelte [hunk 3]
```diff
@@ -98,12 +103,29 @@
...
+    if (isValid) {
       issValue = settings.ISS_DOMAIN
     }
   }
+  function onChangeProductType() {
+    productId = ""
+    product = ""
+  }
+  function onChangeProductId() {
```
Comment (lietu) [inline generator/src/routes/+page.svelte:118]:
```suggestion
  }
  
  function onChangeProductType() {
    productId = ""
    product = ""
  }
  
  function onChangeProductId() {
```

File: generator/src/routes/+page.svelte [hunk 4]
```diff
@@ -98,12 +103,29 @@
...
+    if (productType === ProductType.PREMADE) {
+      product = premadeProducts.find((product) => product.id === productId)?.product!
+    }
+  }
+
+  $: isValid, onChangeValidSignature(), clearError()
+  $: productType, onChangeProductType(), clearError()
+  $: productId, onChangeProductId(), clearError()
+  $: issValue, clearError()
+  $: signOption, clearError()
```
Comment (lietu) [inline generator/src/routes/+page.svelte:128]: This is all very ugly, please use proper event handlers

File: generator/src/routes/+page.svelte [hunk 5]
```diff
@@ -98,12 +103,29 @@
...
     }
   }
+  function onChangeProductType() {
+    productId = ""
+    product = ""
+  }
+  function onChangeProductId() {
+    if (productType === ProductType.PREMADE) {
+      product = premadeProducts.find((product) => product.id === productId)?.product!
+    }
```
Comment (lietu) [inline generator/src/routes/+page.svelte:121]: This logic seems unclear and the name for this function is not great - the dropdown is now "product id", but there is a separate "product id" field. Just seeing "onChangeProductId()" does not in any way to me hint that this has anything to do with selecting a premade product from the dropdown

You should probably call the dropdown "premade product" or something and when it is changed pick the values from it. Also I would rather use "name" as the unique identifier for those entries in this case.

File: generator/src/routes/+page.svelte (hunk 4 above)
Comment (lietu) [reply generator/src/routes/+page.svelte:128]: as in, `<input on:change={onChangeFoo()}>` and `clearError()` in those event handlers

File: generator/src/routes/+page.svelte [hunk 6]
```diff
@@ -131,19 +153,26 @@
...
+              bind:value={product}
+            />
           {:else}
             <FormSelectGroup
               name="product"
               label="Product"
               placeholder="Type a product"
-              options={data.options}
+              options={premadeProducts}
+              bind:value={productId}
```
Comment (lietu) [inline generator/src/routes/+page.svelte:175]: This is now bound to the same value as the input below. It should not be.

**Comment by meandrewandyou** [comment]
Something's wrong with layout on the small resolutions
white_gutter

**Comment by meandrewandyou** [comment]
When dropdown is active, selected option has text cursor, which I believe tells us that we can highlight and edit the text. But we can not. Not sure if it easily fixable here. 
blinking_thing

**Comment by meandrewandyou** [comment]
Oh, we kinda can edit it, but on the single `backspace` whole string gets deleted.
delete

**Comment by lietu** [comment]
I think 320px wide screen is probably the smallest we will ever encounter so below that it doesn't really matter, but until then it's probably a good idea that it wouldn't explode

**Review by lietu — APPROVED** [review]
I can't quickly spot anything more wrong with this

**Pull request merged**